compiler/testData/diagnostics/tests/smartCasts/castchecks/castInTryWithCatch.fir.kt

```diff
@@ -0,0 +1,32 @@
+fun castInTry(s: Any) {
+    try {
+        s as String // Potential cast exception
+    } catch (e: Exception) {
+        s.length // shouldn't be resolved
+    } finally {
+        s.length // shouldn't be resolved
+    }
+    s.length // shouldn't be resolved
+}
+
+fun castInTryAndCatch(s: Any) {
+    try {
+        s as String // Potential cast exception
+    } catch (e: Exception) {
+        s as String // Potential cast exception
+    } finally {
+        s.length // shouldn't be resolved
+    }
+    s.length // shouldn't be resolved
```

Comment:
Ah, this is different from what we discussed at the doc: cast in `catch`. I updated the comment in the test accordingly. Then, the current behavior is conservative (and same as FE 1.0).

Thanks for the suggestion, but I wonder it's already implemented that way? : https://github.com/JetBrains/kotlin/blob/master/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt#L669 Not sure how/where the info in that union node is used.

Nonetheless, I have a different thought about using `union`. Let me use an example of `String?` type variable with null assignment (and indeed, some variants with that scheme popped up as regression quite often).
```
var s: String?
try {
  s = ""
} finally {
  s = null
}
s.length // should be nullable!
```
`merge(flow(tryMainBlockExit), ...)` will have `s -> [exactType: String]`, whereas `flow(finallyExitNode)` will have `s -> [exactType: String?]` if we record type change from null assignment; if not, the flow won't have any type change, and `s` remains as `String`, resulting in unsound smartcast. Anyway, union of those two ends up with `s -> [exactType: {String, String?}]`, since `union`'s type merge operation is `and`, which accumulates the types from incoming flows, and then use them with `intersect`. `intersect(String, String?) => String`, which, again, led to unsound smartcast.

The problem of using `union` is, if we switch the assignments like this:
```
var s: String?
try {
  s = null
} finally {
  s = ""
}
s.length // can be smartcast
```
we won't see any difference: lhs and rhs of `union` are just exchanged. Even though it will be smartcast, it's just a luck.

Okay, going back to the original example:
```
fun castInTryAndCatch(s: Any) {
    try {
        s as String // Potential cast exception
    } catch (e: Exception) {
        s as String // Potential cast exception
    } finally {
        s.length // shouldn't be resolved
    }
    s.length // should be smartcast
}
```
Incoming flows to `finally` would be: `join(flow(tryExpressionEnter), flow(tryMainExit), flow(catchExit))`, which are `s -> Any, s -> String`, and its type `or` is common super type, `Any`. That's why `s.length` in `finally` shouldn't be resolved.

With my changes, the flow at the 2nd `s.length` is same as `finally` block, so no smartcast. With your suggestion of using `union`, well, that's helpful in this case, but as described with other examples, it still miss some cases.

Thus, what I proposed was path-aware flow. `flow(tryExpressionEnter)`, which corresponds to an uncaught exception path, where flow has `s -> Any`, is the culprit that makes smartcast in the last `s.length` infeasible. If we can filter out that path, similar to path-aware property initialization info, we can become more precise again after `try` expression. Note that, we still don't need fixed-point iteration.

Reply:
Yes, I see the problem. Ideally we should save information about reassignments in flow, so it can be properly used in `join` and `union`. But for now I think it's ok to leave this as is (I just need to check your branch on modularized test)